sdk/core/azure-core/src/http/winhttp/win_http_websockets.cpp

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/core/http/http.hpp"
+#include "azure/core/http/policies/policy.hpp"
+#include "azure/core/http/transport.hpp"
+#include "azure/core/http/websockets/win_http_websockets_transport.hpp"
+#include "azure/core/internal/diagnostics/log.hpp"
+#include "azure/core/platform.hpp"
+
+#if defined(AZ_PLATFORM_POSIX)
+#include <poll.h> // for poll()
+#include <sys/socket.h> // for socket shutdown
+#elif defined(AZ_PLATFORM_WINDOWS)
+#if !defined(WIN32_LEAN_AND_MEAN)
+#define WIN32_LEAN_AND_MEAN
+#endif
+#if !defined(NOMINMAX)
+#define NOMINMAX
+#endif
+#include <winapifamily.h>
+#include <winsock2.h> // for WSAPoll();
+#endif
+#include <shared_mutex>
+
+namespace Azure { namespace Core { namespace Http { namespace WebSockets {
+
+  void WinHttpWebSocketTransport::OnUpgradedConnection(
+      Azure::Core::Http::_detail::unique_HINTERNET const& requestHandle)
+  {
+    // Convert the request handle into a WebSocket handle for us to use later.
+    m_socketHandle = Azure::Core::Http::_detail::unique_HINTERNET(
+        WinHttpWebSocketCompleteUpgrade(requestHandle.get(), 0),
+        Azure::Core::Http::_detail::HINTERNET_deleter{});
+    if (!m_socketHandle)
+    {
+      GetErrorAndThrow("Error Upgrading HttpRequest handle to WebSocket handle.");
+    }
+  }
+
+  std::unique_ptr<Azure::Core::Http::RawResponse> WinHttpWebSocketTransport::Send(
+      Azure::Core::Http::Request& request,
+      Azure::Core::Context const& context)
+  {
+    return WinHttpTransport::Send(request, context);
+  }
+
+  /**
+   * @brief  Close the WebSocket cleanly.
+   */
+  void WinHttpWebSocketTransport::NativeClose() { m_socketHandle.reset(); }
```

Comment:
Why do you reset and not close?

Reply:
The way you release a pointer managed by `std::unique_ptr` is to call the `reset()` method, which frees up the underlying object.

In this case, `std::unique_ptr::reset()` calls `HINTERNET_deleter::operator()` which calls `WinHttpCloseHandle`.